src/Extensions/Compatability/Logger/ULSLogging.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Omex.Extensions.Compatability.Logger
+{
+	/// <summary>
+	/// Unified logging, allows registering for logging events sent by
+	/// code implementations and to raise logging events
+	/// </summary>
+	public static class ULSLogging
+	{
+
+		/// <summary>
+		/// Log a trace
+		/// </summary>
+		/// <param name="eventId">tagid (uniqueid) of the trace</param>
+		/// <param name="category">logging category</param>
+		/// <param name="level">logging level</param>
+		/// <param name="message">message to log</param>
+		/// <param name="parameters">additional parameters</param>
+		public static void LogTraceTag(EventId eventId, Category category, Level level, string message, params object[] parameters) =>
+			GetLogger(category)?.Log(ConvertLevel(level), eventId, message, parameters);
+
+
+		/// <summary>
+		/// Report an exception
+		/// </summary>
+		/// <param name="eventId">tag</param>
+		/// <param name="category">category</param>
+		/// <param name="exception">exception</param>
+		/// <param name="message">message</param>
+		/// <param name="parameters">message format parameters</param>
+		public static void ReportExceptionTag(EventId eventId, Category category, Exception exception, string message, params object[] parameters) =>
+			GetLogger(category)?.LogError(eventId, exception, message, parameters);
+
+
+		internal static void Initialize(ILoggerFactory loggerFactory)
+		{
+			s_factory = loggerFactory;
+			s_loggersDictionary = new ConcurrentDictionary<string, ILogger>();
+		}
+
+
+		private static ILogger? GetLogger(Category category) =>
+			s_loggersDictionary != null && s_factory != null
```

Comment:
>s_loggersDictionary != null && s_factory != null [](start = 3, length = 48)

It seems like the goal here is really to check that the class has been initialised. To make the intention clearer, a variable with this info might be best, along with logging indicating exactly what is wrong

Reply:
Without those checks, code won't build, because `s_loggersDictionary` and  `s_factory` are nullable, another option would be to have a method that will return a tuple of no-nullable types, but I might be even more confusing, and about logging, I'm also not sure where to write failure since if it's not initialized there are no logs, but proper exceptions with explanation would be nice

Reply:
Looks like there's no great solution here unfortunately

---
In reply to: [362484743](https://github.com/microsoft/Omex/pull/153#discussion_r362484743) [](ancestors = 362484743)

Reply:
If it's not initialized and it's not a valid use case, it means the contract has been violated, so you can throw an exception in this case and return a non nullable ILogger.

Reply:
Since it's static there might be a case when logger called before initialization, I'm not sure what would be the right action in this case and also tempted to throw an exception, but since this code added for compatibility with old code, I would like it to create fewer problems.